[enhancement] Temporarily store and send data

devkapilbansal opened the pull request:
Closes #29

devkapilbansal commented:
Blocked by #25

nemesisdesign commented on README.rst:
```diff
@@ -31,6 +31,7 @@ UCI configuration options must go in ``/etc/config/monitoring``.
 
 - ``monitored_interfaces``: interfaces that needs to be monitored, defaults to ``*`` for all interfaces.
 - ``interval``: time after which device data should be sent to server, defaults to ``300``
+- ``required_memory``: free memory required to save data temporarily, defaults to ``0.05``
```
is it free or available memory?

devkapilbansal replied:
Free memory

nemesisdesign replied:
there's a problem with this: on modern systems it will always look like the free memory is almost nonexistent.
Linux allocates almost the entire memory to the operating system, ready to be used if needed.
That's why on modern systems you will see that there's also another metric called "available memory".
We should try to use available memory and use free memory only if missing.
Otherwise, we risk this work to be uneffective, because the daemon will think the memory is full and will not store the data.

devkapilbansal replied:
My bad it is available memory. I mentioned free here as that memory used can be freed when needed :thinking:

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -31,6 +32,25 @@ fi
 
 URL="$BASE_URL/api/v1/monitoring/device/$UUID/?key=$KEY"
 
+tmp_dir="/tmp/monitoring"
```
what about `/tmp/openwisp/monitoring`?

devkapilbansal replied:
I will change that

nemesisdesign reviewed (commented):
I see you're saving the data, but when are you going to send it?

devkapilbansal commented:
> I see you're saving the data, but when are you going to send it?

I will send the data once post request is successful #31

nemesisdesign reviewed (changes requested):
There's also a conflict to resolve.

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -31,6 +32,25 @@ fi
+
+check_available_memory(){
+	total=$(free | grep Mem | awk '{print $2}')
+	available=$(free | grep Mem | awk '{print $7}')
```
I take available memory here

feckert replied:
You could use `ubus call system info` to get the information in one call

```
root@st-dev-07 ~ # ubus call system info
{
        "localtime": 1626076905,
        "uptime": 1443394,
        "load": [
                1760,
                1728,
                64
        ],
        "memory": {
                "total": 4117020672,
                "free": 3917463552,
                "shared": 2035712,
                "buffered": 33484800,
                "available": 3958448128,
                "cached": 62803968
        },
        "swap": {
                "total": 0,
                "free": 0
        }
}
```

devkapilbansal replied:
Done

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -31,6 +32,25 @@ fi
 
 URL="$BASE_URL/api/v1/monitoring/device/$UUID/?key=$KEY"
 
+tmp_dir="/tmp/openwisp/monitoring"
```
use uppercase var name please

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -9,6 +9,7 @@ while [ -n "$1" ]; do
 		--verify_ssl) export verify_ssl="$2"; shift;;
 		--interval) export INTERVAL="$2"; shift;;
 		--monitored_interfaces) export MONITORED_INTERFACES=$2; shift;;
+		--required_memory) export required_percent=$2; shift;;
```
use upepr case var name please

nemesisdesign reviewed (changes requested):
Good progress, see my comments below.

I still don't see the code which sends the data to OpenWISP Monitoring, I know the API endpoints to receive the data there not ready yet, but you should start writing the logic which periodically sends it (you can then leave a stub for the POST request which sends it).

BTW: please keep in mind that the sending and the deletion of old data should happen only if the agent can communicate with the server, so probably it should happen if and after the current data is sent successfully, then for each point sent successfully the related data is deleted, iterating on each data point until all are sent and deleted.

devkapilbansal commented:
> Good progress, see my comments below.
> 
> I still don't see the code which sends the data to OpenWISP Monitoring, I know the API endpoints to receive the data there not ready yet, but you should start writing the logic which periodically sends it (you can then leave a stub for the POST request which sends it).
> 
> BTW: please keep in mind that the sending and the deletion of old data should happen only if the agent can communicate with the server, so probably it should happen if and after the current data is sent successfully, then for each point sent successfully the related data is deleted, iterating on each data point until all are sent and deleted.

Taking care of this in #31

nemesisdesign reviewed (changes requested):
> > Good progress, see my comments below.
> > I still don't see the code which sends the data to OpenWISP Monitoring, I know the API endpoints to receive the data there not ready yet, but you should start writing the logic which periodically sends it (you can then leave a stub for the POST request which sends it).
> > BTW: please keep in mind that the sending and the deletion of old data should happen only if the agent can communicate with the server, so probably it should happen if and after the current data is sent successfully, then for each point sent successfully the related data is deleted, iterating on each data point until all are sent and deleted.
> 
> Taking care of this in #31

Since we need both funcionalities (collecting data and sending) to be implemented in order to validate if this code really works, there's no reason to delay the implementation of the sending and work on it in a separate pull request, so please start working on it in this PR.
We can keep commits separate once we squash them, but definitely working on this in two PRs is only going to add overhead.

devkapilbansal commented:
> Since we need both funcionalities (collecting data and sending) to be implemented in order to validate if this code really works, there's no reason to delay the implementation of the sending and work on it in a separate pull request, so please start working on it in this PR.
> We can keep commits separate once we squash them, but definitely working on this in two PRs is only going to add overhead.

Sure, I will work on the sending part in this PR only

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,126 @@ collect_data(){
+        [ "$verbose_mode" = "1" ] && logger -s "Data saved temporarily" \
+                                            -p daemon.info
+    fi
+    # get process id of container sending data
+    pid=$(ps | grep openwisp_monitoring | grep uuid | awk '{print $1}')
+    kill -SIGUSR1 "$pid"
+    sleep "$INTERVAL"
```
Sending SIGUSR1 signal once data is collected

devkapilbansal reviewed (commented):
If I am not wrong, this is what we need :point_down:

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,126 @@ collect_data(){
+  MONITORED_INTERFACES=${MONITORED_INTERFACES:-*}
+  save_data
+else
+  trap send_data USR1
```
Trapping send_data to ensure data is sent as soon as it is collected

devkapilbansal reviewed (commented):
If I am not wrong, this is what we need :point_down:

nemesisdesign commented on README.rst:
```diff
@@ -40,6 +40,8 @@ UCI configuration options must go in ``/etc/config/monitoring``.
 - ``verbose_mode``: can be enabled (set to ``1``) to ease `debugging <#debugging>`__ in case of issues, defaults to ``0`` (disabled).
+- ``required_memory``: available memory required to save data temporarily, defaults to ``0.05``.
+- ``max_retries``: number of time the service should try to send data to server in case of failure, defaults to ``5``.
 
```
I think we need to add some more info in the README to explain how the 2 processes work.

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,139 @@ collect_data(){
+    data="$(collect_data)"
+    memory_available="$(check_available_memory)"
+    if [ "$memory_available" != "0" ]; then
+        file_name="$(date +'%d-%m-%Y_%H:%M:%S').000000.json" # extra zeroes are added for nanoseconds precision
```
Made this change for openwisp-monitoring module.

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+    data="$(collect_data)"
+    memory_available="$(check_available_memory)"
+    if [ "$memory_available" != "0" ]; then
+        file_name="$(date -u +'%d-%m-%Y_%H:%M:%S').000000.json" # extra zeroes are added for nanoseconds precision
```
please move the comment above the line of code to avoid the line being too long

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+        file_name="$(date -u +'%d-%m-%Y_%H:%M:%S').000000.json" # extra zeroes are added for nanoseconds precision
+        # make directory
+        mkdir -p "$TMP_DIR"
+        #save data with file_name
```
always leave a space after the comment

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+}
+
+save_data() {
+  # Remove double quotes from interfaces
```
if you're starting comments with lowercase letters be consistent and ensure the style of all comments is consistent

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -26,40 +30,23 @@ if [ "$VERSION" -eq "1" ]; then
-MONITORED_INTERFACES=$(echo "$MONITORED_INTERFACES" | tr -d '"')
+check_available_memory(){
+  total=$(ubus call system info | grep -m1 total | tr -cd 0-9)
+  available=$(ubus call system info | grep available | tr -cd 0-9)
```
can you use jsonfilter to get this? It should be available by default on OpenWRT and should result in being cleaner.

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -1,6 +1,7 @@
 #!/bin/sh
 
 VERSION=0 #default value of version
+COLLECT=0
```
wouldn't it be more readable if we had something like `MODE="collect"` and `MODE="send"`?

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+
+  while true
+  do
+    data="$(collect_data)"
```
why are you collecting the data before checking for the memory availability?
If the data is large and the memory is already at full capacity, isn't there's a chance the agent my crash here? In that case it would not be as robust as intended wouldn't it?

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+      url="$URL&time=$filename"
+      data=$(cat "$file")
+      # retry sending data in case of failure
+      MAX_RETRY=${MAX_RETRY:-5}
```
2 observations here:

- isn't this value a constant which doesn't change in the loop?
  if so, why define it inside loop, which is executed many times? Can't this be defined just once at the beginning of the function?
- I think it can be renamed to `MAX_RETRIES`

Keep in mind the same suggestion for any other variable which doesn't need to be in the loop, I see a few different ones here so please double check here and in any other loop.

Please also keep in mind that uppercase is generally used for constants which are defined at the beginning of the program and used in different places. For anything else defined only in specific functions, please use lowercase naming and flag them as [local variables](https://tldp.org/LDP/abs/html/localvar.html).

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+      [ "$(echo "$TMP_DIR"/*.json | awk '{print $2}')" ] || url="$url&current=true"
+      while true
+      do
+        FAILURES=$((FAILURES+1))
```
why are you adding 1 here, before any failure?
Is this `FAILURES` variable actually a counter of how many attempts have been made? In that case I would name it `attempts`.

devkapilbansal replied:
Changing in to increase after failure

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+        if [ "$response_code" = "200" ]; then
+          [ "$verbose_mode" = "1" ] && logger -s "Data sent successfully." \
+                                              -p daemon.info
+          #Removed saved data
```
please always leave a space from the beginning of the comment and keep the first case consistent with the other comments. Applies to any other comment.

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+        response_code=$($CURL_COMMAND -H "Content-Type: application/json" \
+          -d "$data" \
+          -v "$url")
+        if [ "$response_code" = "200" ]; then
```
we not use -eq instead of = for consistency with other sections? Let's choose one way of doing it and let's stick to it, however I personally prefer to use `==` instead of `=` because the single equal sign can be confused with assignment.

Applies to all other boolean evaluations.

devkapilbansal replied:
`==` is not defined in strict POSIX. That's why I used single `=` as suggested by `shellcheck`.

Changing it to -eq where integer comparisions are made for consistency.

nemesisdesign replied:
can we use -eq here too?

devkapilbansal replied:
Here, I use `=` beause in case of failure, error message will be there along with `response_code` which will be string. This is changed to `-eq` in #63 where error message is instead stored in a file.

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+          break
+        else
+          TIME=$((BASE_SECONDS * 2 **(FAILURES-1)))
+          [ "$verbose_mode" = "1" ] && logger -s "Data not sent successfully. Retrying after $TIME seconds" \
```
I'd use "Retrying in $TIME seconds"

nemesisdesign commented on README.rst:
```diff
@@ -40,6 +40,18 @@ UCI configuration options must go in ``/etc/config/monitoring``.
 - ``monitored_interfaces``: interfaces that needs to be monitored, defaults to ``*`` for all interfaces.
 - ``interval``: time after which device data should be sent to server, defaults to ``300``.
 - ``verbose_mode``: can be enabled (set to ``1``) to ease `debugging <#debugging>`__ in case of issues, defaults to ``0`` (disabled).
+- ``required_memory``: available memory required to save data temporarily, defaults to ``0.05``.
```
what is 0.05, a percentage? Please make it explicit.

nemesisdesign commented on README.rst:
```diff
@@ -40,6 +40,18 @@ UCI configuration options must go in ``/etc/config/monitoring``.
 - ``interval``: time after which device data should be sent to server, defaults to ``300``.
 - ``verbose_mode``: can be enabled (set to ``1``) to ease `debugging <#debugging>`__ in case of issues, defaults to ``0`` (disabled).
+- ``required_memory``: available memory required to save data temporarily, defaults to ``0.05``.
+- ``max_retries``: number of time the service should try to send data to server in case of failure, defaults to ``5``.
```
suggestion:

maximum number of retries in case of failures to send data to the server, defaults to ``5``.
in case the maximum number of retries is reached .... (link the section below so the user knows what happens in this case).

nemesisdesign commented on README.rst:
```diff
@@ -40,6 +40,18 @@ UCI configuration options must go in ``/etc/config/monitoring``.
+- ``max_retries``: number of time the service should try to send data to server in case of failure, defaults to ``5``.
+
+OpenWISP Monitoring Agent
+-------------------------
```
This title is not very informative of what is being explained here.

We want to explain that this agent works with 2 processes, one for collecting and one for sending.

We could name this section "Collecting vs sending".

nemesisdesign commented on README.rst:
```diff
@@ -40,6 +40,18 @@ UCI configuration options must go in ``/etc/config/monitoring``.
+**SIGUSR1** signals are used to instantly send the data when collected. However, the service will keep trying
+to send data periodically.
+
+**Note:-** It may be possible that the server is down for a long time, then sending data may take time then the interval after which data is collected creating a gap in data points. To avoid this, two services are used.
```
Citing the server as being down is not a great example, the most common case is the OpenWRT device simply not being able to send the data for network connectivity issues.

I think we could divide this section further and explain what happens in the most important cases, eg:

1st heading: collecting vs sending

Intro: this agent uses 2 different processes, one for sending data and one for collecting it.
This allows to handle failures in sending the data in a more flexible way because old data stored during network connectivity issues can be sent at the same time while new data is being collected: if old data has piled up for sending and takes several minutes to be uploaded, new data is collected without being delayed by the sending.

Subheading 1: Collect mode

In this mode the agent is in charge of collecting data ...
The data is collected only if ...
If there is not enough memory ...
Once data is collected, a signal is sent to the other agent so that ...

Subheading 2: Send mode

In this mode the agent .....
If sending fails .... explain retries...
If all attempts fail ...

nemesisdesign commented on README.rst:
```diff
@@ -96,7 +108,6 @@ you will need to select the *openwisp-monitoring* variant and *netjson-monitorin
     ./scripts/feeds update -a
     ./scripts/feeds install -a
     make menuconfig
-    # go to Base system, then select rpcd
```
are you sure this is needed and are we sure this section of explainin how to select the packages with makemenu config is needed at all? The explanations above are enough in most cases, this seems redundant to me.

devkapilbansal replied:
This was not, that's why I removed it :sweat_smile:

devkapilbansal replied:
I added this section to make it consistent with `**openwisp-config**

nemesisdesign replied:
lol, I think it may be a good idea to remove it from there too then :grin: 

hey what do you think @nickberry17 @feckert, do you think this is useful at all?

devkapilbansal replied:
Would it be better to replace the section with a GIF?

![OpenWISP Package Compiling](https://user-images.githubusercontent.com/42766576/131209183-60b8308d-a283-43a7-bb3c-25fa3db9cfe5.gif)

nemesisdesign replied:
Yes, is always good to have something visual :+1: 
pelase upload the gif to a media branch (to keep the main branch lightweight).

nickberry17 replied:
> Would it be better to replace the section with a GIF?
> 
> ![OpenWISP Package Compiling](https://user-images.githubusercontent.com/42766576/131209183-60b8308d-a283-43a7-bb3c-25fa3db9cfe5.gif)

I like the idea of a GIF, but it add a note saying it's not necessary to select all of those packages; only the one you need.

nemesisdesign reviewed (changes requested):
Good progress @devkapilbansal, please see my comments below.

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -78,24 +65,140 @@ collect_data(){
+        [ "$verbose_mode" = "1" ] && logger -s "Data saved temporarily" \
+                                            -p daemon.info
+    fi
+    # get process id of container sending data
```
:grimacing: Fixed now

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -26,76 +29,184 @@ if [ "$VERSION" -eq "1" ]; then
+                                          -p daemon.info
+    fi
+    # get process id of the process sending data 
+    pid=$(pgrep -f "openwisp_monitoring.*uuid")
```
We also don't need to evaluate process id each time but in starting collect data will starts first before send data :thinking:

nemesisdesign replied:
but the other process may be restarted

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -26,76 +29,184 @@ if [ "$VERSION" -eq "1" ]; then
+check_available_memory(){
+  total=$(ubus call system info | jsonfilter -e '@.memory.total')
+  available=$(ubus call system info | jsonfilter -e '@.memory.available')
+  required=$(echo - |awk -v percent="$REQUIRED_PERCENT" -v total="$total" '{printf("%.f",percent*total)}')
```
Is there a way to improve this too? I tried searching for one but end in futile.

nemesisdesign replied:
at the moment I don't have ideas on how to further improve this.

devkapilbansal reviewed (commented):
Made some minor improvements in the agent as suggested above and will add a commit for documentation soon.
I have some doubts here :point_down:

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -167,38 +161,52 @@ send_data() {
   VERIFY_SSL=${VERIFY_SSL:-0}
   set_url_and_curl && send_data
+else
+  logger -s "Please specify correct mode" \
```
instead of logging an error, I suggest just echoing to standard error a message like "The supplied mode is invalid, only collect and send are accepted."

nemesisdesign commented on README.rst:
```diff
@@ -40,6 +40,46 @@ UCI configuration options must go in ``/etc/config/monitoring``.
+
+We use two procd services in `monitoring agent <https://github.com/openwisp/openwrt-openwisp-monitoring/blob/master/openwrt-openwisp-monitoring/files/monitoring.agent>`_, one for collecting the data and other for sending the data.
+
+This helps handle failure in sending the data in more flexible way. Old data saved during network connectivity issues can be sent while new data is being collected. If old data has piled up and takes several minutes to be uploaded, new data will be collected without waiting for the sending to complete.
```
shouldn't it be "helps to handle"?

nemesisdesign commented on README.rst:
```diff
@@ -40,6 +40,46 @@ UCI configuration options must go in ``/etc/config/monitoring``.
+If data is sent successfully, then the data file will be deleted and agent will look for another file.
+
+**SIGUSR1** signals are used to instantly send the data when collected. However, the service will keep trying
+to send data periodically.
```
@nickberry17 your review of this text would be super helpful!

nemesisdesign reviewed (commented):
Looks good, but I need to test it still (couldn't test it because it seems the device currently available in our test system doesn't have the agent, I left some hints in the chat channel).

@feckert @nickberry17 your input in this PR will be much appreciated :pray:

nemesisdesign reviewed (changes requested):
Good progress @devkapilbansal, I tested it quickly and it seems to me that the basic goals are reached, so :+1:  :clap: for that! 

Here's the issues I found during testing.
I haven't looked into OpenWISP Monitoring yet.

### Rule of silence is broken

**Rule of Silence: When a program has nothing surprising to say, it should say nothing.**
http://www.catb.org/~esr/writings/taoup/html/ch01s06.html#id2878450

Looking at the log, I see:

```
Fri Aug 13 10:30:24 2021 daemon.info openwisp_monitoring: SIGUSR1 received! Sending data
Fri Aug 13 10:35:25 2021 daemon.info openwisp_monitoring: SIGUSR1 received! Sending data
Fri Aug 13 10:40:26 2021 daemon.info openwisp_monitoring: SIGUSR1 received! Sending data
```

This would fill any log collector with useless data, let's avoid this unless in verbose mode.

The same is valid when there's something wrong: we do not need to log every failed request when not in verbose mode, we can simply log the first failure and let the user know the sending is not working, then when the sending works again we can log a line saying sending is working again and that's it.

### Name of files stored could be shortened to save memory

I see the filenames are something like:

`13-08-2021_01:39:29.000000.json`

But I think the last part is unnecessarily eating the memory of the device.

Can you add `.000000` programmatically instead of having to use it in the file name?

Then, I think we can live without `.json`, isn't it?

devkapilbansal commented:
> ```
> Fri Aug 13 10:30:24 2021 daemon.info openwisp_monitoring: SIGUSR1 received! Sending data
> Fri Aug 13 10:35:25 2021 daemon.info openwisp_monitoring: SIGUSR1 received! Sending data
> Fri Aug 13 10:40:26 2021 daemon.info openwisp_monitoring: SIGUSR1 received! Sending data
> ```
> 
> This would fill any log collector with useless data, let's avoid this unless in verbose mode.

:grimacing: Fixed now

> 
> The same is valid when there's something wrong: we do not need to log every failed request when not in verbose mode, we can simply log the first failure and let the user know the sending is not working, then when the sending works again we can log a line saying sending is working again and that's it.

:thinking: Would it be like maintaining a variable to know when the request start failing and when the process is normal again. Like, logging only when the value of variable change

> ### Name of files stored could be shortened to save memory
> I see the filenames are something like:
> 
> `13-08-2021_01:39:29.000000.json`
> 
> But I think the last part is unnecessarily eating the memory of the device.
> 
> Can you add `.000000` programmatically instead of having to use it in the file name?

I added it in URL directly. :+1: 

> 
> Then, I think we can live without `.json`, isn't it?

I added `.json` previously to distinguish data files from any other file that we may save in future. Removed now.

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -26,76 +29,179 @@ if [ "$VERSION" -eq "1" ]; then
-CURL_COMMAND="curl -s -w "%{http_code}""
-[ "$VERIFY_SSL" = "0" ] && CURL_COMMAND="$CURL_COMMAND -k"
-[ "$verbose_mode" = "1" ] && CURL_COMMAND="$CURL_COMMAND -v"
+  if [ "$available" -ge "$required" ]; then
+    echo "1"
+  else
+    [ "$verbose_mode" -eq "1" ] && logger -s "Not enough memory available" \
+                                          -p daemon.err
+    echo "0"
```
I think this should be changed too.
0 should be returned if case of success, else we should return 1

devkapilbansal replied:
Changed

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -26,76 +29,179 @@ if [ "$VERSION" -eq "1" ]; then
+      fi
+      basefilename=${file##*/}
+      filename=${basefilename%.*}
+      # extra zeroes are added for nanoseconds precision
+      url="$URL&time=$filename.000000"
```
Added zeroes while sending data in url directly.

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -26,76 +29,180 @@ if [ "$VERSION" -eq "1" ]; then
+  exit 1
+fi
+
+if [ "$MODE" = "collect" ]; then
```
can we use -eq here too please?

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -26,76 +29,180 @@ if [ "$VERSION" -eq "1" ]; then
+  # remove double quotes from interfaces 
+  MONITORED_INTERFACES=$(echo "$MONITORED_INTERFACES" | tr -d '"')
+  save_data
+elif [ "$MODE" = "send" ]; then
```
can we use -eq here too please?

nemesisdesign reviewed (changes requested):
It's working a lot better, see my comments below, we should be near to merge this.

nemesisdesign reviewed (changes requested):
I just noticed two issues.

1. error message is shown more than once in non verbose mode:

Thu Aug 19 02:03:08 2021 daemon.err openwisp_monitoring: Data not sent successfully. Response code is 000 Error message is  Run with verbose mode to find more.
Thu Aug 19 02:08:08 2021 daemon.err openwisp_monitoring: Data not sent successfully. Response code is 000 Error message is  Run with verbose mode to find more.
Thu Aug 19 02:13:09 2021 daemon.err openwisp_monitoring: Data not sent successfully. Response code is 000 Error message is  Run with verbose mode to find more.

2. the error message is ... 

I think this:

`Response code is 000 Error message is  Run with verbose mode to find more.`

Would be better written like this:

`Response code is "000". Error message is "". Run with verbose mode to find more.`

devkapilbansal commented:
> I just noticed two issues.
> 
> 1. error message is shown more than once in non verbose mode:
> 
> ```
> Thu Aug 19 02:03:08 2021 daemon.err openwisp_monitoring: Data not sent successfully. Response code is 000 Error message is  Run with verbose mode to find more.
> Thu Aug 19 02:08:08 2021 daemon.err openwisp_monitoring: Data not sent successfully. Response code is 000 Error message is  Run with verbose mode to find more.
> Thu Aug 19 02:13:09 2021 daemon.err openwisp_monitoring: Data not sent successfully. Response code is 000 Error message is  Run with verbose mode to find more.
> ```
> 
> 1. the error message is ...
> 
> I think this:
> 
> `Response code is 000 Error message is Run with verbose mode to find more.`
> 
> Would be better written like this:
> 
> `Response code is "000". Error message is "". Run with verbose mode to find more.`

This error message is coming from #63 that I included with this to test. Will fix it there

nemesisdesign commented:
> > I just noticed two issues.
> > 
> > 1. error message is shown more than once in non verbose mode:
> > 
> > ```
> > Thu Aug 19 02:03:08 2021 daemon.err openwisp_monitoring: Data not sent successfully. Response code is 000 Error message is  Run with verbose mode to find more.
> > Thu Aug 19 02:08:08 2021 daemon.err openwisp_monitoring: Data not sent successfully. Response code is 000 Error message is  Run with verbose mode to find more.
> > Thu Aug 19 02:13:09 2021 daemon.err openwisp_monitoring: Data not sent successfully. Response code is 000 Error message is  Run with verbose mode to find more.

I was thinking, since in openwisp-config I noticed there's a log line every time there's a failure, I think this is fine.
It may be more useful you focus on other things.

nemesisdesign commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -10,7 +10,10 @@ while [ -n "$1" ]; do
     --interval) export INTERVAL="$2"; shift;;
     --monitored_interfaces) export MONITORED_INTERFACES="$2"; shift;;
-    --verbose_mode) export verbose_mode=$2; shift;;
+    --verbose_mode) export verbose_mode="$2"; shift;;
```
verbose_mode is still lowercase

devkapilbansal commented on openwrt-openwisp-monitoring/files/monitoring.agent:
```diff
@@ -26,76 +29,178 @@ if [ "$VERSION" -eq "1" ]; then
+          rm "$file"
+          break
+        else
+          TIME=$((BASE_SECONDS * 2 **(failures)))
```
I was trying to get exponent instead of square, i.e. 2^failures. Nvm, I use a cleaner way to do so.

nemesisdesign reviewed (approved):
Thank you very much @nickberry17 @devkapilbansal :+1:
